File: app/src/main/java/maw/org/wedding/map/HotelsInfoWindowAdapter.java
package maw.org.wedding.map;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Marker;
import com.squareup.picasso.Picasso;

import butterknife.ButterKnife;
import maw.org.wedding.R;

public class HotelsInfoWindowAdapter implements GoogleMap.InfoWindowAdapter {

    private View mView;

    public HotelsInfoWindowAdapter(Context context) {
        mView = LayoutInflater.from(context).inflate(R.layout.hotel_marker_layout, null);
    }

    @Override
    public View getInfoWindow(final Marker marker) {

        if (marker.getTitle() != null) {
            ButterKnife.findById(mView, R.id.content).setVisibility(View.VISIBLE);
            TextView hotelNameView = ButterKnife.findById(mView, R.id.hotel_name);
            hotelNameView.setText(marker.getTitle());

            ImageView hotelImageView = ButterKnife.findById(mView, R.id.hotel_image);

            Picasso.with(mView.getContext()).load(marker.getSnippet()).into(hotelImageView);
        }
        else {
            ButterKnife.findById(mView, R.id.content).setVisibility(View.GONE);
        }

        return mView;
    }

    @Override
    public View getInfoContents(Marker marker) {

        if (marker.getTitle() != null) {
            ButterKnife.findById(mView, R.id.content).setVisibility(View.VISIBLE);
            TextView hotelNameView = ButterKnife.findById(mView, R.id.hotel_name);
            hotelNameView.setText(marker.getTitle());

            ImageView hotelImageView = ButterKnife.findById(mView, R.id.hotel_image);
            Picasso.with(mView.getContext()).load(marker.getSnippet()).into(hotelImageView);

        }
        else {
            ButterKnife.findById(mView, R.id.content).setVisibility(View.GONE);
        }

        return mView;
    }
}

File: app/src/main/kotlin/maw/org/wedding/map/HotelsInfoWindowAdapter.kt
package maw.org.wedding.map

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Marker
import com.squareup.picasso.Picasso
import maw.org.wedding.R
import java.util.*

class HotelsInfoWindowAdapter(context: Context) : GoogleMap.InfoWindowAdapter, MarkerViewController {

    private val mView: View

    private val mMarkerViewModelMap = HashMap<String, MarkerViewModel>()

    init {
        mView = LayoutInflater.from(context).inflate(R.layout.hotel_marker_layout, null)
    }

    override fun getInfoWindow(marker: Marker): View? {
        return null

    }

    override fun getInfoContents(marker: Marker): View {

        val markerViewModel = mMarkerViewModelMap[marker.id]

        if (markerViewModel != null) {
            mView.findViewById(R.id.content).visibility = View.VISIBLE
            val hotelNameView = mView.findViewById(R.id.hotel_name) as TextView
            hotelNameView.text = markerViewModel.title

            val hotelImageView = mView.findViewById(R.id.hotel_image) as ImageView

            if (markerViewModel.imageUrls.size > 0) {
                Picasso.with(mView.context).load(markerViewModel.imageUrls[0]).into(hotelImageView)
            }
        } else {
            mView.findViewById(R.id.content).visibility = View.GONE
        }

        return mView
    }

    override fun updateViewModelForMarker(marker: Marker, markerViewModel: MarkerViewModel) {
        mMarkerViewModelMap.put(marker.id, markerViewModel)
        marker.showInfoWindow()
    }

    override fun markerHasData(markerID: String): Boolean {
        return mMarkerViewModelMap.containsKey(markerID)
    }

    override fun getMarkerViewModel(id: String): MarkerViewModel {
        return mMarkerViewModelMap[id]!!
    }
}
